patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs

```
@@ -0,0 +1,184 @@
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(ModItem modItem, int stack = 1)
+		{
+			_recipe.AddIngredient(modItem, stack);
+
+			return this;
+		}
+
+
+		/// <summary>Adds a recipe group ingredient to this recipe with the given RecipeGroup name and stack size. Vanilla recipe groups consist of "Wood", "IronBar", "PresurePlate", "Sand", and "Fragment".</summary>
+		/// <param name="recipeGroup">The name.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException"></exception>
+		public RecipeBuilder Requires(string recipeGroup, int stack = 1)
+		{
+			_recipe.AddRecipeGroup(recipeGroup, stack);
+
+			return this;
+		}
+
+
+		/// <summary>
+		/// Adds one or many required crafting station(s) with the given tile type(s) to the recipe being built.
+		/// Ex.:
+		/// <example>At(TileID.WorkBenches, TileID.Anvils)</example>
+		/// </summary>
+		/// <param name="tileTypes"></param>
+		/// <returns></returns>
+		public RecipeBuilder At(params int[] tileTypes)
+		{
+			for (int i = 0; i < tileTypes.Length; i++)
+				_recipe.AddTile(tileTypes[i]);
+
+			return this;
+		}
+
+
+		public RecipeBuilder RequiresLava() => NeedVar(recipe => recipe.needLava = true);
+
+		public RecipeBuilder RequiresHoney() => NeedVar(recipe => recipe.needHoney = true);
+
+		public RecipeBuilder RequiresWater() => NeedVar(recipe => recipe.needWater = true);
+
+		public RecipeBuilder RequiresSnowBiome() => NeedVar(recipe => recipe.needSnowBiome = true);
+
+
+		private RecipeBuilder NeedVar(Action<ModRecipe> need)
+		{
+			need(_recipe);
+
+			return this;
+		}
+
+
+		/// <summary>Adds this recipe to the game. Call this after you have finished setting the result, ingredients, etc.</summary>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">A recipe without any result has been added.</exception>
+		public void Build() => _recipe.AddRecipe();
```

You should have a 'finalized' or 'added' bool on this builder and check against it in any of the method calls, the builder is 'done' once Build() is called.

---

Thats a lot of hand-holding. Do we need to throw an error every time as well ?